frontend/wasm/src/module/build_ir.rs

```diff
@@ -203,19 +206,39 @@ fn build_data_segments(
     module_builder: &mut ModuleBuilder,
     diagnostics: &DiagnosticsHandler,
 ) -> WasmResult<()> {
+    // First, collect all data segments into ResolvedDataSegment structures
+    let mut resolved_segments = Vec::new();
```

Comment:
*Nit:* I'd use `SmallVec` for things like this, especially since most modules we'll compile will have < 3 segments. The less temporary heap allocation we do during compilation, the better.

Reply:
Thanks! Done.